### DISCUSSION TITLE
More like a question, rather than issue

### DISCUSSION THREAD

**Discussion by hngts** [ajaxorg/ace#5444] [Q&A] [answered]
Hello there. I am using Ace editor for quite a while already, and I don't have any issue with so far. Ace is the best thing I've seen being made in Javascript. 

Here is my dilemma. 

- I have made new custom `mode-funcdown.js`mode for my little project called `Funcdown`. 
- It's some sort of advanced markup genetator with it's own syntax. 
- Here it is live, rendered with Ace, with new custom `mode-funcdown.js`  [Funcdown Playground](https://hardcoder.xyz/?mkp=fncd#)
- Here is the permalink to the custom mode : [hngts/mode-funcdown.js](https://github.com/hngts/funcdown-highlighters/blob/17f8579892d28b4d229fc1fcd998349973fb8ed5/mode-funcdown.js) 

The problem is that I don't do that much of advanced .js nor the tests, etc .. and I can confirm that the custom mode I made works ok .. as I saw that there are tests.js with all modes in main branch .... 

If that mode above is OK, where should I commit it ? Which branch ? src ? src-no-conflict ?  
That's the dilemma. 

I need some guidance regarding adding new ace modes. 
Thanks in advance.

**Comment by InspiredGuy** [comment, marked as answer]
Hey @hngts , thank you for reaching out about a potential contribution. You can find a helpful instruction on Ace website https://ace.c9.io/#nav=higlighter, but here's the condensed and complete (more or less) version:
- create a mode and highlight rules (in your case seems like this is done already, but in a single file - we recommend splitting them up)
- generate tokens for syntax highlighting tests https://github.com/ajaxorg/ace/blob/master/src/mode/_test/Readme.md
- add tests for the mode specific behaviours, if your mode has custom behaviours and not just inherits some other mode with minor tweaks
- add mode to the modelist `src/ext/modelist.js`
- add mode type declaration to `ace-modes.d.ts`
- add an example document for the demo page

There are several examples of PRs which add modes, here are couple of them: https://github.com/ajaxorg/ace/pull/5356, https://github.com/ajaxorg/ace/pull/5342. Your contribution would be welcome as well!

**Reply by hngts** [reply]
> generate tokens for syntax highlighting tests https://github.com/ajaxorg/ace/blob/master/src/mode/_test/Readme.md

Is there any way I can write tests without node ? The thing is that I never wrote a single test in js so far.

I have quite decent experience with php/sql primarely and vanilla js and css. For more than 15 years. I have created a lots of websites and smaller webshops. I know how to write bash, lua .. a bit of Rust etc. .. but .. never worked with **ts** nor anything other than browser-based js engine(s) which includes node.js as well. This might seem weird a bit, but that's the case. 

I don't even know ace.js api well enough (to be honnest), although I am using Ace quite a lots onto my website. 

Also, I have chosen to copy and paste _define "ace/mode/folding/coffee"_   ... at the very beginning of the mode as it seemed like a smallest and efficient one I could find that would complement folding on all wrappers/brace types, and I don't have a clue whether some basic mode for custom mode is needed at all ?!? .. in first place. 

Conclusion is that I don't think I am competent enough to finish all that on the list by my self. Not in some shorter period of time.

**Comment by InspiredGuy** [comment]
I have moved your issue from `ace-builds` to a discussion in `ace` repository, since `ace-builds` is a repo dedicated to distribution of ready Ace builds. This repo is about the editor itself, which also includes language modes.

**Reply by hngts** [reply]
Thanks. It is embarrasing from me, but I have just saw (at a time of writing this reply) your response. Will get into it.
